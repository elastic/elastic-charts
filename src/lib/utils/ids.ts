import { iso, Newtype } from 'newtype-ts';

export interface GroupId extends Newtype<{ readonly GroupId: unique symbol }, string> { }
export interface AxisId extends Newtype<{ readonly AxisId: unique symbol }, string> { }
export interface SpecId extends Newtype<{ readonly SpecId: unique symbol }, string> { }
export interface AnnotationId extends Newtype<{ readonly AnnotationId: unique symbol }, string> { }

const groupIdIso = iso<GroupId>();
const axisIdIso = iso<AxisId>();
const specIdIso = iso<SpecId>();
const annotationIdIso = iso<AnnotationId>();

export function getGroupId(id: string): GroupId {
  return groupIdIso.wrap(id);
}
export function getAxisId(id: string): AxisId {
  return axisIdIso.wrap(id);
}
export function getSpecId(id: string): SpecId {
  return specIdIso.wrap(id);
}
export function getAnnotationId(id: string): AnnotationId {
  return annotationIdIso.wrap(id);
}
