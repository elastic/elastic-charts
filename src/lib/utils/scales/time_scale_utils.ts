import { DateTime, DurationObject, Interval } from 'luxon';

export function computeMinTimeInterval(data: number[]): DurationObject[] {
  const valuesLength = data.length;
  if (valuesLength <= 0) {
    return [{ millisecond: 0 }];
  }
  if (valuesLength === 1) {
    return [{ millisecond: 0 }];
  }
  const sortedValues = data.slice().sort();
  let i;
  const intervals: DurationObject[] = [];

  for (i = 1; i < valuesLength - 1; i++) {
    const current = DateTime.fromMillis(sortedValues[i]);
    const next = DateTime.fromMillis(sortedValues[i + 1]);
    const interval = Interval.fromDateTimes(current, next);
    const yearInterval = interval.toDuration('year').years;
    if (Number.isInteger(yearInterval)) {
      intervals.push({ year: yearInterval });
      continue;
    }
    const monthInterval = interval.toDuration('month').months;

    if (Number.isInteger(monthInterval)) {
      intervals.push({ month: monthInterval });
      continue;
    }
    const weekInterval = interval.toDuration('week').weeks;
    if (Number.isInteger(weekInterval)) {
      intervals.push({ week: weekInterval });
      continue;
    }
    const dayInterval = interval.toDuration('day').days;
    if (Number.isInteger(dayInterval)) {
      intervals.push({ day: dayInterval });
      continue;
    }
    const millisInterval = interval.toDuration('milliseconds').milliseconds;
    intervals.push({
      milliseconds: millisInterval,
    });
  }
  return intervals;
}
