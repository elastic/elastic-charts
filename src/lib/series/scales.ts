import { GroupId } from '../utils/ids';
import {
  createContinuousScale,
  createOrdinalScale,
  Scale,
  ScaleType,
} from '../utils/scales/scales';
import { XDomain } from './domains/x_domain';
import { YDomain } from './domains/y_domain';
import { FormattedDataSeries } from './series';

/**
 * Count the max number of bars in cluster value.
 * Doesn't take in consideration areas, lines or points.
 * @param stacked all the stacked formatted dataseries
 * @param nonStacked all the non-stacked formatted dataseries
 */
export function countBarsInCluster(
  stacked: FormattedDataSeries[],
  nonStacked: FormattedDataSeries[],
): {
  nonStackedBarsInCluster: number;
  stackedBarsInCluster: number;
  totalBarsInCluster: number;
} {
  // along x axis, we count one "space" per bar series.
  // we ignore the points, areas, lines as they are
  // aligned with the x value and doesn't occupy space
  const nonStackedBarsInCluster = nonStacked.reduce((acc, ns) => {
    return acc + ns.counts.barSeries;
  }, 0);
  // count stacked bars groups as 1 per group
  const stackedBarsInCluster = stacked.reduce((acc, ns) => {
    return acc + (ns.counts.barSeries > 0 ? 1 : 0);
  }, 0);
  const totalBarsInCluster = nonStackedBarsInCluster + stackedBarsInCluster;
  return {
    nonStackedBarsInCluster,
    stackedBarsInCluster,
    totalBarsInCluster,
  };
}

/**
 * Compute the x scale used to align geometries to the x axis.
 * @param xDomain the x domain
 * @param totalBarsInCluster the total number of grouped series
 * @param axisLength the length of the x axis
 */
export function computeXScale(
  xDomain: XDomain,
  totalBarsInCluster: number,
  minRange: number,
  maxRange: number,
): Scale {
  const { scaleType, minInterval, domain, isBandScale } = xDomain;
  const rangeDiff = Math.abs(maxRange - minRange);
  const isInverse = maxRange < minRange;
  if (scaleType === ScaleType.Ordinal) {
    const dividend = totalBarsInCluster > 0 ? totalBarsInCluster : 1;
    const bandwidth = rangeDiff / (domain.length * dividend);
    return createOrdinalScale(domain, minRange, maxRange, 0, bandwidth);
  } else {
    if (isBandScale) {
      const intervalCount = (domain[1] - domain[0]) / minInterval;
      const bandwidth = rangeDiff / (intervalCount + 1);
      const start = isInverse ? minRange - bandwidth : minRange;
      const end = isInverse ? maxRange : maxRange - bandwidth;
      return createContinuousScale(
        scaleType,
        domain,
        start,
        end,
        bandwidth / totalBarsInCluster,
        false,
        minInterval,
      );
    } else {
      return createContinuousScale(
        scaleType,
        domain,
        minRange,
        maxRange,
        0,
        undefined,
        minInterval,
      );
    }
  }
}

/**
 * Compute the y scales, one per groupId for the y axis.
 * @param yDomains the y domains
 * @param axisLength the axisLength of the y axis
 */
export function computeYScales(
  yDomains: YDomain[],
  minRange: number,
  maxRange: number,
): Map<GroupId, Scale> {
  const yScales: Map<GroupId, Scale> = new Map();

  yDomains.forEach((yDomain) => {
    const yScale = createContinuousScale(yDomain.scaleType, yDomain.domain, minRange, maxRange);
    yScales.set(yDomain.groupId, yScale);
  });

  return yScales;
}
