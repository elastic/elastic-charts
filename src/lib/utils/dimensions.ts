import { AxisTicksDimensions } from '../axes/axis_utils';
import { AxisSpec, Position } from '../series/specs';
import { Theme } from '../themes/theme';
import { AxisId } from './ids';

export interface Dimensions {
  top: number;
  left: number;
  width: number;
  height: number;
}

export interface Margins {
  top: number;
  bottom: number;
  left: number;
  right: number;
}

/**
 * Compute the chart dimension padding the parent dimension by the specified set of axis
 * @param parentDimensions the parent dimension
 * @param axisDimensions the axis dimensions
 * @param axisSpecs the axis specs
 */
export function computeChartDimensions(
  parentDimensions: Dimensions,
  chartTheme: Theme,
  axisDimensions: Map<AxisId, AxisTicksDimensions>,
  axisSpecs: Map<AxisId, AxisSpec>,
  showLegend: boolean,
  legendPosition?: Position,
): Dimensions {
  const chartMargins = chartTheme.chart.margins;
  const chartPaddings = chartTheme.chart.paddings;
  const legendStyle = chartTheme.legend;
  const { titleFontSize, titlePadding } = chartTheme.axes;

  const axisTitleHeight = titleFontSize + titlePadding;

  let vLeftAxisSpecWidth = 0;
  let vRightAxisSpecWidth = 0;
  let hTopAxisSpecHeight = 0;
  let hBottomAxisSpecHeight = 0;

  axisDimensions.forEach(({ maxLabelBboxWidth = 0, maxLabelBboxHeight = 0 }, id) => {
    const axisSpec = axisSpecs.get(id);
    if (!axisSpec) {
      return;
    }
    const { position, tickSize, tickPadding } = axisSpec;
    switch (position) {
      case Position.Top:
        hTopAxisSpecHeight +=
          maxLabelBboxHeight + tickSize + tickPadding + chartMargins.top + axisTitleHeight;
        break;
      case Position.Bottom:
        hBottomAxisSpecHeight +=
          maxLabelBboxHeight + tickSize + tickPadding + chartMargins.bottom + axisTitleHeight;
        break;
      case Position.Left:
        vLeftAxisSpecWidth +=
          maxLabelBboxWidth + tickSize + tickPadding + chartMargins.left + axisTitleHeight;
        break;
      case Position.Right:
        vRightAxisSpecWidth +=
          maxLabelBboxWidth + tickSize + tickPadding + chartMargins.right + axisTitleHeight;
        break;
    }
  });
  // const hMargins = chartMargins.left + chartMargins.right;
  const chartWidth = parentDimensions.width - vLeftAxisSpecWidth - vRightAxisSpecWidth;
  const chartHeight = parentDimensions.height - hTopAxisSpecHeight - hBottomAxisSpecHeight;
  let vMargin = 0;
  if (hTopAxisSpecHeight === 0) {
    vMargin += chartMargins.top;
  }
  if (hBottomAxisSpecHeight === 0) {
    vMargin += chartMargins.bottom;
  }
  let hMargin = 0;
  if (vLeftAxisSpecWidth === 0) {
    hMargin += chartMargins.left;
  }
  if (vRightAxisSpecWidth === 0) {
    hMargin += chartMargins.right;
  }
  let legendTopMargin = 0;
  let legendLeftMargin = 0;
  if (showLegend) {
    switch (legendPosition) {
      case Position.Right:
        hMargin += legendStyle.verticalWidth;
        break;
      case Position.Left:
        hMargin += legendStyle.verticalWidth;
        legendLeftMargin = legendStyle.verticalWidth;
        break;
      case Position.Top:
        vMargin += legendStyle.horizontalHeight;
        legendTopMargin = legendStyle.horizontalHeight;
        break;
      case Position.Bottom:
        vMargin += legendStyle.horizontalHeight;
        break;
    }
  }
  let top = 0;
  let left = 0;
  if (hTopAxisSpecHeight === 0) {
    top = chartMargins.top + chartPaddings.top + legendTopMargin;
  } else {
    top = hTopAxisSpecHeight + chartPaddings.top + legendTopMargin;
  }
  if (vLeftAxisSpecWidth === 0) {
    left = chartMargins.left + chartPaddings.left + legendLeftMargin;
  } else {
    left = vLeftAxisSpecWidth + chartPaddings.left + legendLeftMargin;
  }
  return {
    top,
    left,
    width: chartWidth - hMargin - chartPaddings.left - chartPaddings.right,
    height: chartHeight - vMargin - chartPaddings.top - chartPaddings.bottom,
  };
}
