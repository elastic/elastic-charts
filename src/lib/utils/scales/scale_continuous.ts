import { bisectLeft } from 'd3-array';
import { scaleLinear, scaleLog, scaleSqrt, scaleUtc } from 'd3-scale';
import { DateTime } from 'luxon';
import { clamp } from '../commons';
import { ScaleContinuousType, ScaleType, Scale } from './scales';

const SCALES = {
  [ScaleType.Linear]: scaleLinear,
  [ScaleType.Log]: scaleLog,
  [ScaleType.Sqrt]: scaleSqrt,
  [ScaleType.Time]: scaleUtc,
};

export function limitToMin(value: number, positive: boolean) {
  if (value === 0) {
    return positive ? 1 : -1;
  }
  return value;
}
/**
 * As log(0) = -Infinite, a log scale domain must be strictly-positive
 * or strictly-negative; the domain must not include or cross zero value.
 * We need to limit the domain scale to the right value on all possible cases.
 * @param domain the domain to limit
 */
export function limitLogScaleDomain(domain: any[]) {
  if (domain[0] === 0) {
    if (domain[1] > 0) {
      return [1, domain[1]];
    } else if (domain[1] < 0) {
      return [-1, domain[1]];
    } else {
      return [1, 1];
    }
  }
  if (domain[1] === 0) {
    if (domain[0] > 0) {
      return [domain[0], 1];
    } else if (domain[0] < 0) {
      return [domain[0], -1];
    } else {
      return [1, 1];
    }
  }
  if (domain[0] < 0 && domain[1] > 0) {
    const isD0Min = Math.abs(domain[1]) - Math.abs(domain[0]) >= 0;
    if (isD0Min) {
      return [1, domain[1]];
    } else {
      return [domain[0], -1];
    }
  }
  if (domain[0] > 0 && domain[1] < 0) {
    const isD0Max = Math.abs(domain[0]) - Math.abs(domain[1]) >= 0;
    if (isD0Max) {
      return [domain[0], 1];
    } else {
      return [-1, domain[1]];
    }
  }
  return domain;
}

export class ScaleContinuous implements Scale {
  readonly bandwidth: number;
  readonly totalBarsInCluster: number;
  readonly bandwidthPadding: number;
  readonly minInterval: number;
  readonly step: number;
  readonly type: ScaleType;
  readonly domain: any[];
  readonly range: number[];
  readonly isInverted: boolean;
  readonly tickValues: number[];
  readonly timeZone: string;
  readonly barsPadding: number;
  private readonly d3Scale: any;

  constructor(
    type: ScaleContinuousType,
    domain: any[],
    range: [number, number],
    /**
     * The desidered bandwidth for a linear band scale.
     * @default 0
     */
    bandwidth: number = 0,
    /**
     * The min interval computed on the XDomain. Not available for yDomains.
     * @default 0
     */
    minInterval: number = 0,
    /**
     * A time zone identifier. Can be any IANA zone supported by he host environment,
     * or a fixed-offset name of the form 'utc+3', or the strings 'local' or 'utc'.
     * @default 'utc'
     */
    timeZone: string = 'utc',
    /**
     * The number of bars in the cluster. Used to correctly compute scales when
     * using padding between bars.
     * @default 1
     */
    totalBarsInCluster: number = 1,
    /**
     * The proportion of the range that is reserved for blank space between bands
     * A number between 0 and 1.
     * @default 0
     */
    barsPadding: number = 0,
  ) {
    this.d3Scale = SCALES[type]();
    if (type === ScaleType.Log) {
      this.domain = limitLogScaleDomain(domain);
      this.d3Scale.domain(this.domain);
    } else {
      this.domain = domain;
      this.d3Scale.domain(domain);
    }
    const safeBarPadding = clamp(barsPadding, 0, 1);
    this.barsPadding = safeBarPadding;
    this.bandwidth = bandwidth * (1 - safeBarPadding);
    this.bandwidthPadding = bandwidth * safeBarPadding;
    this.d3Scale.range(range);
    this.step = 0;
    this.type = type;
    this.range = range;
    this.minInterval = minInterval;
    this.isInverted = this.domain[0] > this.domain[1];
    this.timeZone = timeZone;
    this.totalBarsInCluster = totalBarsInCluster;
    if (type === ScaleType.Time) {
      const startDomain = DateTime.fromMillis(this.domain[0], { zone: this.timeZone });
      const endDomain = DateTime.fromMillis(this.domain[1], { zone: this.timeZone });
      const offset = startDomain.offset;
      const shiftedDomainMin = startDomain.plus({ minutes: offset }).toMillis();
      const shiftedDomainMax = endDomain.plus({ minutes: offset }).toMillis();
      const tzShiftedScale = scaleUtc().domain([shiftedDomainMin, shiftedDomainMax]);

      const rawTicks = tzShiftedScale.ticks();
      const timePerTick = (shiftedDomainMax - shiftedDomainMin) / rawTicks.length;
      const hasHourTicks = timePerTick < 1000 * 60 * 60 * 12;

      this.tickValues = rawTicks.map((d: Date) => {
        const currentDateTime = DateTime.fromJSDate(d, { zone: this.timeZone });
        const currentOffset = hasHourTicks ? offset : currentDateTime.offset;
        return currentDateTime.minus({ minutes: currentOffset }).toMillis();
      });
    } else {
      if (this.minInterval > 0) {
        const intervalCount = Math.floor((this.domain[1] - this.domain[0]) / this.minInterval);
        this.tickValues = new Array(intervalCount + 1).fill(0).map((d, i) => {
          return this.domain[0] + i * this.minInterval;
        });
      } else {
        this.tickValues = this.d3Scale.ticks();
      }
    }
  }

  scale(value: any) {
    return this.d3Scale(value) + (this.bandwidthPadding / 2) * this.totalBarsInCluster;
  }

  ticks() {
    return this.tickValues;
  }
  invert(value: number) {
    let invertedValue = this.d3Scale.invert(value);
    if (this.type === ScaleType.Time) {
      invertedValue = DateTime.fromJSDate(invertedValue).toMillis();
    }
    return invertedValue;
  }
  invertWithStep(value: number, data: number[]): any {
    const invertedValue = this.invert(value - this.bandwidth / 2);
    const leftIndex = bisectLeft(data, invertedValue);
    if (leftIndex === 0) {
      // is equal or less than the first value
      const prevValue1 = data[leftIndex];
      if (data.length === 0) {
        return prevValue1;
      }
      const nextValue1 = data[leftIndex + 1];
      const nextDiff1 = Math.abs(nextValue1 - invertedValue);
      const prevDiff1 = Math.abs(invertedValue - prevValue1);
      if (nextDiff1 < prevDiff1) {
        return nextValue1;
      }
      return prevValue1;
    }
    if (leftIndex === data.length) {
      return data[leftIndex - 1];
    }
    const nextValue = data[leftIndex];
    const prevValue = data[leftIndex - 1];
    const nextDiff = Math.abs(nextValue - invertedValue);
    const prevDiff = Math.abs(invertedValue - prevValue);
    if (nextDiff <= prevDiff) {
      return nextValue;
    }
    return prevValue;
  }
  isSingleValue() {
    if (this.domain.length < 2) {
      return true;
    }

    const min = this.domain[0];
    const max = this.domain[this.domain.length - 1];
    return max === min;
  }
}

export function isContinuousScale(scale: Scale): scale is ScaleContinuous {
  return scale.type !== ScaleType.Ordinal;
}

export function isLogarithmicScale(scale: Scale) {
  return scale.type === ScaleType.Log;
}
