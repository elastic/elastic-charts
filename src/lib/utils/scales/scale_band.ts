import { scaleBand, scaleQuantize, ScaleQuantize } from 'd3-scale';
import { StepType } from './scale_continuous';
import { ScaleType } from './scales';
import { Scale } from './scales';

export class ScaleBand implements Scale {
  readonly bandwidth: number;
  readonly step: number;
  readonly type: ScaleType;
  readonly domain: any[];
  readonly range: number[];
  readonly isInverted: boolean;
  readonly invertedScale: ScaleQuantize<number>;
  readonly minInterval: number;
  private readonly d3Scale: any;

  constructor(
    domain: any[],
    range: [number, number],
    padding?: [number, number],
    round?: boolean,
    overrideBandwidth?: number,
  ) {
    this.type = ScaleType.Ordinal;
    this.d3Scale = scaleBand();
    this.d3Scale.domain(domain);
    this.d3Scale.range(range);
    if (padding) {
      this.d3Scale.paddingInner(padding[0]);
      this.d3Scale.paddingOuter(padding[1]);
    }
    if (round) {
      this.d3Scale.round(round);
    }
    this.bandwidth = this.d3Scale.bandwidth() || 0;
    this.step = this.d3Scale.step();
    this.domain = this.d3Scale.domain();
    this.range = range.slice();
    if (overrideBandwidth) {
      this.bandwidth = overrideBandwidth;
    }
    // TO FIX: we are assiming that it's ordered
    this.isInverted = this.domain[0] > this.domain[1];
    this.invertedScale = scaleQuantize()
      .domain(range)
      .range(this.domain);
    this.minInterval = 0;
  }

  scale(value: any) {
    return this.d3Scale(value);
  }

  ticks() {
    return this.domain;
  }
  invert(value: any) {
    return this.invertedScale(value);
  }
  invertWithStep(value: any, stepType?: StepType) {
    return this.invertedScale(value);
  }
}

export function isOrdinalScale(scale: Scale): scale is ScaleBand {
  return scale.type === ScaleType.Ordinal;
}
