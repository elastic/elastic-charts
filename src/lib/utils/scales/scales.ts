import { StepType } from './scale_continuous';

export interface Scale {
  domain: any[];
  range: number[];
  ticks: () => any[];
  scale: (value: any) => number;
  invert: (value: number) => any;
  invertWithStep: (value: number, stepType?: StepType) => any;
  bandwidth: number;
  minInterval: number;
  type: ScaleType;
  isInverted: boolean;
  barsPadding: number;
}
export type ScaleFunction = (value: any) => number;

/**
 * The scale type
 */
export const enum ScaleType {
  Linear = 'linear',
  Ordinal = 'ordinal',
  Log = 'log',
  Sqrt = 'sqrt',
  Time = 'time',
}

export interface ScaleConfig {
  accessor: (value: any) => any;
  domain: any[];
  type: ScaleType;
  clamp?: boolean;
}

export type ScaleContinuousType =
  | ScaleType.Linear
  | ScaleType.Sqrt
  | ScaleType.Log
  | ScaleType.Time;
export type ScaleOrdinalType = ScaleType.Ordinal;
export type ScaleTypes = ScaleContinuousType | ScaleOrdinalType;
