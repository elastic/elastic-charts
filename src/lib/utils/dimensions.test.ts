import { AxisTicksDimensions } from '../axes/axis_utils';
import { AxisSpec, Position } from '../series/specs';
import { LIGHT_THEME } from '../themes/light_theme';
import { LegendStyle } from '../themes/theme';
import { computeChartDimensions, Margins } from './dimensions';
import { AxisId, getAxisId, getGroupId } from './ids';
import { ScaleType } from './scales/scales';

describe('Computed chart dimensions', () => {
  const parentDim = {
    width: 100,
    height: 100,
    top: 0,
    left: 0,
  };
  const chartMargins: Margins = {
    left: 10,
    right: 10,
    top: 10,
    bottom: 10,
  };
  const chartPaddings: Margins = {
    left: 10,
    right: 10,
    top: 10,
    bottom: 10,
  };

  const axis1Dims: AxisTicksDimensions = {
    axisScaleType: ScaleType.Linear,
    axisScaleDomain: [0, 1],
    tickValues: [0, 1],
    tickLabels: ['first', 'second'],
    maxLabelBboxWidth: 10,
    maxLabelBboxHeight: 10,
    maxLabelTextWidth: 10,
    maxLabelTextHeight: 10,
  };
  const axisLeftSpec: AxisSpec = {
    id: getAxisId('axis_1'),
    groupId: getGroupId('group_1'),
    hide: false,
    showOverlappingTicks: false,
    showOverlappingLabels: false,
    position: Position.Left,
    tickSize: 10,
    tickPadding: 10,
    tickFormat: (value: any) => {
      return `${value}`;
    },
  };
  const legend: LegendStyle = {
    verticalWidth: 10,
    horizontalHeight: 10,
  };
  const showLegend = false;
  const defaultTheme = LIGHT_THEME;
  const chartTheme = {
    ...defaultTheme,
    chartMargins,
    chartPaddings,
    axes: {
      ...defaultTheme.axes,
    },
    ...legend,
  };
  chartTheme.axes.axisTitleStyle.fontSize = 10;
  chartTheme.axes.axisTitleStyle.padding = 10;
  test('should be equal to parent dimension with no axis minus margins', () => {
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    const chartDimensions = computeChartDimensions(
      parentDim,
      chartTheme,
      axisDims,
      axisSpecs,
      showLegend,
    );
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a left axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxWidth|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from left
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), axisLeftSpec);
    const chartDimensions = computeChartDimensions(
      parentDim,
      chartTheme,
      axisDims,
      axisSpecs,
      showLegend,
    );
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a right axis', () => {
    // |padding|tickSize|tickPadding|maxLabelBBoxWidth|titlePadding|titleFontSize\margin|
    // \10|10|10|10|10|10|10| = 70px from right
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), { ...axisLeftSpec, position: Position.Right });
    const chartDimensions = computeChartDimensions(
      parentDim,
      chartTheme,
      axisDims,
      axisSpecs,
      showLegend,
    );
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a top axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxHeight|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from top
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), {
      ...axisLeftSpec,
      position: Position.Top,
    });
    const chartDimensions = computeChartDimensions(
      parentDim,
      chartTheme,
      axisDims,
      axisSpecs,
      showLegend,
    );
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a bottom axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxHeight|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from bottom
    const axisDims = new Map<AxisId, AxisTicksDimensions>();
    const axisSpecs = new Map<AxisId, AxisSpec>();
    axisDims.set(getAxisId('axis_1'), axis1Dims);
    axisSpecs.set(getAxisId('axis_1'), {
      ...axisLeftSpec,
      position: Position.Bottom,
    });
    const chartDimensions = computeChartDimensions(
      parentDim,
      chartTheme,
      axisDims,
      axisSpecs,
      showLegend,
    );
    expect(chartDimensions).toMatchSnapshot();
  });
});
