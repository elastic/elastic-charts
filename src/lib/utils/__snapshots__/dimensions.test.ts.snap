// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Computed chart dimensions should be equal to parent dimension with no axis minus margins 1`] = `
Object {
  "height": 60,
  "left": 20,
  "top": 20,
  "width": 60,
}
`;

exports[`Computed chart dimensions should be padded by a bottom axis 1`] = `
Object {
  "height": 10,
  "left": 20,
  "top": 20,
  "width": 60,
}
`;

exports[`Computed chart dimensions should be padded by a left axis 1`] = `
Object {
  "height": 60,
  "left": 70,
  "top": 20,
  "width": 10,
}
`;

exports[`Computed chart dimensions should be padded by a right axis 1`] = `
Object {
  "height": 60,
  "left": 20,
  "top": 20,
  "width": 10,
}
`;

exports[`Computed chart dimensions should be padded by a top axis 1`] = `
Object {
  "height": 10,
  "left": 20,
  "top": 70,
  "width": 60,
}
`;
